File: appendix02/course-tracker-building-webapp/course-tracker-final/src/main/java/com/manning/sbip/a02/controller/CourseController.java
package com.manning.sbip.a02.controller;

import com.manning.sbip.a02.model.Course;
import com.manning.sbip.a02.service.CourseService;
import org.springframework.stereotype.Controller;
import org.springframework.ui.Model;
import org.springframework.validation.BindingResult;
import org.springframework.web.bind.annotation.*;

import javax.validation.Valid;
import java.util.Collections;
import java.util.List;

@Controller
public class CourseController {

    private final CourseService courseService;

    public CourseController(CourseService courseService) {
        this.courseService = courseService;
    }

    @GetMapping("/")
    public String index() {
        return "redirect:/index";
    }

    @GetMapping("/index")
    public String index(Model model) {
        List<Course> courseList = (List<Course>) courseService.findAllCourses();
        model.addAttribute("courses", courseList.isEmpty() ? Collections.EMPTY_LIST : courseList);
        return "index";
    }

    @GetMapping("/addcourse")
    public String showAddCourseForm(Course course) {
        return "add-course";
    }

    @PostMapping("/addcourse")
    public String addCourse(@Valid Course course, BindingResult result, Model model){
        if (result.hasErrors()) {
            return "add-course";
        }
        model.addAttribute("courses", courseService.createCourse(course));
        return "redirect:/index";
    }

    @GetMapping("/update/{id}")
    public String showUpdateCourseForm(@PathVariable("id") int id, Model model) {
        model.addAttribute("course", courseService.findCourseById(id).get());
        return "update-course";
    }

    @PutMapping("/update/{id}")
    public String updateCourse(@PathVariable("id") int id, @Valid Course course, BindingResult result, Model model) {
        if (result.hasErrors()) {
            course.setId(id);
            return "update-course";
        }
        model.addAttribute("courses", courseService.updateCourse(course));
        return "redirect:/index";
    }

    @DeleteMapping("/delete/{id}")
    public String deleteCourse(@PathVariable("id") int id, Model model) {
        model.addAttribute("courses", courseService.deleteCourseById(id));
        return "redirect:/index";
    }
}

File: ch10/Kotlin/course-tracker-kotlin-app/src/main/kotlin/com/manning/sbip/ch10/controller/CourseController.kt
package com.manning.sbip.ch10.controller

import com.manning.sbip.ch10.model.Course
import com.manning.sbip.ch10.service.CourseService
import org.springframework.stereotype.Controller
import org.springframework.ui.Model
import org.springframework.ui.set
import org.springframework.validation.BindingResult
import org.springframework.web.bind.annotation.*
import java.util.*
import javax.validation.Valid

@Controller
class CourseController(private val courseService: CourseService) {

    @GetMapping("/")
    fun index(): String {
        return "redirect:/index"
    }

    @GetMapping("/index")
    fun index(model: Model): String {
        val courseList = courseService.findAllCourses() as List<Course>
        model["courses"] = if (courseList.isEmpty()) Collections.EMPTY_LIST else courseList
        return "index"
    }

    @GetMapping("/addcourse")
    fun showAddCourseForm(course: Course): String {
        return "add-course"
    }

    @PostMapping("/addcourse")
    fun addCourse(course: @Valid Course, result: BindingResult, model: Model): String {
        if (result.hasErrors()) {
            return "add-course"
        }
        courseService.createCourse(course)
        model["courses"] = courseService.findAllCourses();
        return "redirect:/index"
    }

    @GetMapping("/update/{id}")
    fun showUpdateCourseForm(@PathVariable("id") id: Long, model: Model): String {
        model["course"] = courseService.findCourseById(id)
        return "update-course"
    }

    @PutMapping("/update/{id}")
    fun updateCourse(@PathVariable("id") id: Long, course: @Valid Course, result: BindingResult, model: Model): String {
        if (result.hasErrors()) {
            course.id = id
            return "update-course"
        }
        courseService.updateCourse(id, course)
        model["courses"] = courseService.findAllCourses()
        return "redirect:/index"
    }

    @DeleteMapping("/delete/{id}")
    fun deleteCourse(@PathVariable("id") id: Long, model: Model): String {
        courseService.deleteCourseById(id)
        model["courses"] = courseService.findAllCourses()
        return "redirect:/index"
    }
}